src/Razor/src/Microsoft.AspNetCore.Razor.VSCode/src/CodeActions/RazorFullyQualifiedCodeActionTranslator.ts

```diff
@@ -13,13 +13,44 @@ export class RazorFullyQualifiedCodeActionTranslator implements IRazorCodeAction
     public applyEdit(
         uri: vscode.Uri,
         edit: vscode.TextEdit): [vscode.Uri | undefined, vscode.TextEdit | undefined] {
-        // The edit for this should just translate without additional help.
-        throw new Error('Method not implemented.');
+        // This is kind of wrong. We're manually trimming whitespace and adjusting
+        // for multi-line edits that're provided by O#. Right now, we do not support multi-line edits
+        // (ex. refactoring code actions) however there are certain supported edits which O# is automatically
+        // formatting for us (ex. FullyQualifiedNamespace) into multiple lines, when it should span a single line.
+        // This is due to how we render our virtual cs files, with fewer levels of indentation to facilitate
+        // appropriate error reporting (if we had additional tabs, then the error squigly would appear offset).
+        //
+        // The ideal solution for this would do something like:
+        // https://github.com/dotnet/aspnetcore-tooling/blob/4c8dbd0beb073e6dcee33d96d174453bf44d938a/
+        // src/Razor/src/Microsoft.AspNetCore.Razor.LanguageServer/Formatting/DefaultRazorFormattingService.cs#L264
+        // however we're going to hold off on that for now as it isn't immediately necessary and we don't
+        // (currently) support any other kind of multi-line edits.
+
+        const newText = edit.newText.trim();
+
+        // The starting and ending range may be equal in the case when we have other items on the same line. Ex:
+        // Render|Tree apple
+        // where `|` is the cursor. We want to ensure we dont't overwrite `apple` in this case with our edit.
+        if (newText !== edit.newText && !edit.range.start.isEqual(edit.range.end)) {
```

Comment:
Why this (`!edit.range.start.isEqual(edit.range.end)`) condition?  When will this be false?

Reply:
Let `|` represent the cursor.

```
RenderTree|
counter++;
```
In this case the `range.start` would be the `R` char of `RenderTree` and the `range.end` would be the last char `e` of `RenderTree` (as the edit is the complete `Microsoft.AspNetCore.Components.RenderTree`).

`!edit.range.start.isEqual(edit.range.end) == true`.

```
RenderTree| apple
counter++;
```
In this case the `range.start` would be the `R` char of `RenderTree` and the `range.end` would be also be the `R` char of `RenderTree` (as the edit is only `Microsoft.AspNetCore.Components.`).

`!edit.range.start.isEqual(edit.range.end) == false`.

In this case we don't need to change the `range.end` as the edit is "in place". We just need to trail the leading whitespace.